cares_reinforcement_learning/algorithm/mbrl/DYNA_SAC.py

```diff
@@ -0,0 +1,292 @@
+import os
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+
+
+class MBRL_DYNA_SAC:
+    """
+    Use the Soft Actor Critic as the Actor Critic framework.
+
+    """
+
+    def __init__(
+        self,
+        actor_network,
+        critic_network,
+        world_network,
+        gamma,
+        tau,
+        action_num,
+        actor_lr,
+        critic_lr,
+        alpha_lr,
+        num_samples,
+        horizon,
+        device,
+    ):
+        self.type = "mbrl"
+        # Switches
+        self.num_samples = num_samples
+        self.horizon = horizon
+        self.action_num = action_num
+        # Other Variables
+        self.gamma = gamma
+        self.tau = tau
+        self.device = device
+        self.batch_size = None
+
+        # this may be called policy_net in other implementations
+        self.actor_net = actor_network.to(device)
+        # this may be called soft_q_net in other implementations
+        self.critic_net = critic_network.to(device)
+        self.target_critic_net = copy.deepcopy(self.critic_net).to(device)
+
+        # Set to initial alpha to 1.0 according to other baselines.
+        self.log_alpha = torch.tensor(np.log(1.0)).to(device)
+        self.log_alpha.requires_grad = True
+        self.target_entropy = -action_num
+
+        # optimizer
+        self.actor_net_optimiser = torch.optim.Adam(
+            self.actor_net.parameters(), lr=actor_lr
+        )
+        self.critic_net_optimiser = torch.optim.Adam(
+            self.critic_net.parameters(), lr=critic_lr
+        )
+        self.log_alpha_optimizer = torch.optim.Adam([self.log_alpha], lr=alpha_lr)
+
+        # World model
+        self.world_model = world_network
+        self.world_model.to(device)
+        self.learn_counter = 0
+        self.policy_update_freq = 1
+
+    @property
+    def alpha(self):
+        """
+        A variatble decide to what extend entropy shoud be valued.
+        """
+        return self.log_alpha.exp()
+
+    # pylint: disable-next=unused-argument to keep the same interface
+    def select_action_from_policy(self, state, evaluation=False, noise_scale=0):
+        """
+        Select a action for executing. It is the only channel that an agent
+        will communicate the the actual environment.
+
+        """
+        # note that when evaluating this algorithm we need to select mu as
+        # action so _, _, action = self.actor_net.sample(state_tensor)
+        self.actor_net.eval()
+        with torch.no_grad():
+            state_tensor = torch.FloatTensor(state).unsqueeze(0).to(self.device)
+            if evaluation is False:
+                (action, _, _) = self.actor_net.sample(state_tensor)
+            else:
+                (_, _, action) = self.actor_net.sample(state_tensor)
+            action = action.cpu().data.numpy().flatten()
+        self.actor_net.train()
+        return action
+
+    def true_train_policy(self, states, actions, rewards, next_states, dones):
```

Comment:
rename - if this is a private part of the training - then make it "_train_policy:

Reply:
It will be exposed to the train_loop.